File: C#/binary-tree-inorder-traversal.cs
// Title: 94. Binary Tree Inorder Traversal
// LeetCode Link: https://leetcode.com/problems/binary-tree-inorder-traversal/


// Definition for a binary tree node.
public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution
{
    public void Traverse(TreeNode root, List<int> nodeVals)
    {
        if (root == null) return;

        Traverse(root.left, nodeVals);
        nodeVals.Add(root.val);
        Traverse(root.right, nodeVals);
    }

    public IList<int> InorderTraversal(TreeNode root)
    {
        IList<int> nodeVals = new List<int>();

        Traverse(root, nodeVals);

        return nodeVals;
    }
}

File: TypeScript/binary-tree-inorder-traversal.ts
// Title: 94. Binary Tree Inorder Traversal
// LeetCode Link: https://leetcode.com/problems/binary-tree-inorder-traversal/

// Definition for a binary tree node.
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function traverse(root: TreeNode | null, nodeVals: number[]): void {
  if (root != null) {
    traverse(root.left, nodeVals);
    nodeVals.push(root.val);
    traverse(root.right, nodeVals);
  }
}

function inorderTraversal(root: TreeNode | null): number[] {
  const nodeVals: number[] = [];

  traverse(root, nodeVals);

  return nodeVals;
}
